python/hsfs/usage.py

```diff
@@ -0,0 +1,255 @@
+        return self.method_counts.get(s, 0)
+
+    def should_sample(self, m):
+        cnt = self.get_count(m)
+        if cnt < 100:
+            return True
+        elif cnt < 1000:
+            return random.random() < 0.1
+        elif cnt < 10000:
+            return random.random() < 0.01
+        else:
+            return random.random() < 0.001
+
+    def _get_method_name(self, m):
+        return m.__module__ + m.__name__
+
+
+logging.basicConfig(stream=sys.__stdout__)
+_logger = logging.getLogger(__name__)
+_logger.setLevel(logging.INFO)
+_conn = http.client.HTTPSConnection("usage.hops.works")
+_executor = concurrent.futures.ThreadPoolExecutor(max_workers=2)
+_env_attr = EnvironmentAttribute()
+_method_counter = MethodCounter()
+_backend_hostname = None
+_backend_version = None
+HOPSWORKS_DIR = join(expanduser("~"), ".hopsworks")
+_USER_ID_FILE = "user_id"
+_is_enabled = os.environ.get("ENABLE_HOPSWORKS_USAGE", default="true").lower() == "true"
+
+
+def enable():
+    global _is_enabled
+    _is_enabled = True
+
+
+def disable():
+    global _is_enabled
+    _is_enabled = False
+
+
+def init_usage(hostname, backend_version):
+    global _backend_hostname, _backend_version, _is_enabled
+    _backend_hostname = hostname
+    _backend_version = backend_version
+    _is_enabled = _is_enabled and _is_target_hostname(hostname)
+
+
+def _is_target_hostname(hostname):
+    # Add "localhost" in the first release for testing.
+    target_hostname = {"c.app.hopsworks.ai", "localhost"}
+    return hostname in target_hostname
+
+
+def _hash_string(input_string):
+    if input_string:
+        hash_object = hashlib.md5()
+        hash_object.update(input_string.encode("utf-8"))
+        hashed_string = hash_object.hexdigest()
+        return hashed_string
+    else:
+        return ""
+
+
+def _create_hopsworks_dir_if_not_exist():
+    if not os.path.exists(HOPSWORKS_DIR):
+        os.makedirs(HOPSWORKS_DIR)
+    return HOPSWORKS_DIR
+
+
+def _generate_user_id():
+    random_uuid = uuid.uuid4()
+    # Convert the UUID to a hexadecimal string and remove hyphens
+    uuid_hex = random_uuid.hex.replace("-", "")
+    short_uuid = "HOPSWORKSID=" + uuid_hex[:16]
+    return short_uuid
+
+
+def _extract_method_name_line_and_file(e):
+    tb_frames = traceback.extract_tb(e.__traceback__)
+
+    tb_result = []
+    for frame in tb_frames[::-1]:
+        method_name = frame.name
+        line_number = frame.lineno
+        file_name = frame.filename.split("/")[-1]
+        tb_result.append(f"{file_name}::{line_number} {method_name}")
+    return "\t".join(tb_result)
+
+
+def method_logger(func):
+    # Disable usage BEFORE import hsfs, return function itself
+    if not _is_enabled:
+        return func
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        # Disable usage AFTER import hsfs, return function itself
+        if not _is_enabled:
+            return func(*args, **kwargs)
+
+        start_time = time.time()
```

Comment:
In Python 3.3, perf_counter() was introduced which is a bit more accurate ([docs](https://docs.python.org/3/library/time.html#time.perf_counter)). Can we use that one instead?